### ISSUE TITLE
! can't be cast to Fn(_) -> _

### ISSUE THREAD

**Issue by Pr0methean** [rust-lang/rust#120153]
I tried this code:

```rust
#[derive(Debug)]
pub struct SharedBufferRng<const WORDS_PER_SEED: usize, const SEEDS_CAPACITY: usize, SourceType: Rng,
FallbackFn: Send + Sync + Clone + Fn() -> SourceType> {
    receiver: Receiver<DefaultableAlignedArray<WORDS_PER_SEED, u64>>,
    calling_thread_fallback: Option<FallbackFn>
}
impl<
        const WORDS_PER_SEED: usize,
        const SEEDS_CAPACITY: usize,
        SourceType: Rng + Send + Debug + 'static,
        FallbackFn: Sync + Send + Clone + Fn() -> SourceType
    > SharedBufferRng<WORDS_PER_SEED, SEEDS_CAPACITY, SourceType, FallbackFn>
    where [(); WORDS_PER_SEED * size_of::<u64>()]:, [(); WORDS_PER_SEED * SEEDS_CAPACITY * size_of::<u64>()]:
{
    /// Creates an RNG that will have a new dedicated thread reading from [source] into a new buffer that's shared with
    /// all clones of this [SharedBufferRng].
    pub fn new(mut source: SourceType, fallback: Option<FallbackFn>) -> Self {/*...*/}
}
#[test]
fn test() {
     let seeder: SharedBufferRng<8, 4, _, _> =
         SharedBufferRng::new(BlockRng64::new(source.clone()), None::<!>);
}
```

I expected to see this happen: the code would compile. `Option<!>` would be a valid type with zero size, and would cast to `Option<T>` for all T, because the `Some` variant would be eliminated during monomorphization. In fact `None` would always be of type `Option<!>` unless and until the compiler found a way for it to become `Some`.

Instead, this happened: it didn't compile until I changed the turbofish to `::<fn(_) -> _>`
### Meta

`rustc --version --verbose`:
```
rustc 1.77.0-nightly (fb5ed726f 2023-12-28)
binary: rustc
commit-hash: fb5ed726f72c6d16c788517c60ec00d4564b9348
commit-date: 2023-12-28
host: x86_64-apple-darwin
release: 1.77.0-nightly
LLVM version: 17.0.6
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by clubby789** [comment]
Can you provide a minimal reproducible example for this issue?

**Comment by Pr0methean** [comment]
Here's one:
```
#![feature(never_type)]

fn maybe_run<T: Fn()>(optional_func: Option<T>) {
    match optional_func {
        Some(func) => func(),
        None => {}
    }
}

fn main() {
    maybe_run::<!>(None)
}
```

**Comment by fmease** [comment]
> Here's one

That one just demonstrates that `!` doesn't impl `FnOnce`

**Comment by fmease** [comment]
`!` is not the bottom type – the subtype of all types. `!` can only be *coerced* to `T` for any type `T`. This coercion isn't transitive: You can't coerce `(!,)` to `(T,)` for any type `T` for example. Same with `Option<!>` and `Option<T>` for any type `T`.

**Comment by fmease** [comment]
If you want to coerce `Option<!>` to `Option<T>` for any type `T`, you'll need to do it manually (“*lift* the coercion”), e.g. via `map`: `o.map(|x| x)`.

**Comment by Pr0methean** [comment]
If `!` isn't the bottom type (as I'd understood it to be), then how is the actual bottom type (one that e.g. won't have any enum variants that include it as a field survive monomorphization) represented?

**Comment by Pr0methean** [comment]
IMO the `!` type won't be doing its job properly until `Vec<!>::len()` is susceptible to constant folding.

**Comment by Pr0methean** [comment]
And this is certainly doable, since for example Kotlin has a `Nothing` type even when compiling for the JVM (which doesn't).

**Comment by kadiwa4** [comment]
`Fn() -> SourceType` (from the first example you posted) is just a trait, with `fn() -> SourceType` only being one of the possible types it could be. So implicitly casting `None::<!>` to `None::<fn() -> SourceType>` – just based on the requirement that the type has to implement `Fn() -> SourceType` – wouldn't make any sense to me.
Especially because an implementation of `Fn(..) -> _ for !` could be added at some point which doesn't sound like a bad idea actually. Is that what you were trying to suggest with this issue?

But maybe what you meant is something like this:
```rust
#![feature(never_type)]

fn maybe_run(optional_func: Option<fn()>) {
    match optional_func {
        Some(func) => func(),
        None => {}
    }
}

fn main() {
    maybe_run(None::<!>)
}
```
In which case fmease's answer applies.

`size_of::<Vec<!>>()` won't be zero anytime soon because you can't currently have specialized structs for specific type parameters. `Vec::<!>::len()` can never return anything because it can't be called. `size_of::<Option<!>>()` is already zero as a compile-time constant because the `Some` variant is deleted as you say.

**Comment by Pr0methean** [comment]
I agree that `impl Fn(_) -> _ for !` is what I meant to suggest, since the first example was designed to monomorphize for both method references and closures with and without captured variables.

**Comment by WaffleLapkin** [comment]
> If `!` isn't the bottom type (as I'd understood it to be), then how is the actual bottom type (one that e.g. when used as a field type in an enum variant, deletes that variant during monomorphization) represented?

`!` isn't a bottom type in the sense that it isn't a **subtype** of all types, but rather merely _coerces_ to all (sized) types. Compiler still uses it for all kinds of optimizations (such as deleting dead code, or removing unused enum variants from the layout).

> And this is certainly doable, since for example Kotlin has a `Nothing` type (used as the return type for methods that are provably non-halting or that always throw an exception) even when compiling for the JVM (which doesn't have a bottom type).

Kotlin and JVM support subtyping, which allows the bottom type to actually be a subtype of all types. Rust doesn't have the luxury of reference semantics, and thus cannot support subtyping to the same degree (we only have subtyping for lifetimes).

> I agree that `impl Fn() -> _ for !` is what I meant to suggest, since the first example was designed to monomorphize for both method references and closures with and without captured variables.

Having such an implementation is not possible, since the `_` is an output parameter (associated type) and thus needs to be determined solely based on the input parameters (`!` in this example). Thus, there can only be 1 implementation, which chooses one specific type as the output type.

Note that you can coerce `!` to a [_function pointer_](https://doc.rust-lang.org/stable/std/primitive.fn.html) (`fn()`, note lowercase `f`) which then implements `Fn()`.
